Frontend/HsTypeChecker.hs

```diff
@@ -519,6 +495,161 @@ rnTyVarsToFcTypes = map rnTyVarToFcType
 rnTmVarToFcTerm :: RnTmVar -> FcTerm
 rnTmVarToFcTerm = FcTmVar . rnTmVarToFcTmVar
 
+-- * Pattern Match Elaboration
+-- ----------------------------------------------------
+
+-- | Suprised all this doesn't exist yet? Is it needed?
+-- | Substitute (in the given pattern) all variables equal to given HsTmVar with second HsTmVar
+substPat :: Eq a => HsTmVar a -> HsTmVar a -> HsPat a -> HsPat a
+substPat v1 v2 (HsPatVar v)
+  | v == v1                     = HsPatVar v2
+  | otherwise                   = HsPatVar v
+substPat v1 v2 (HsPatCons c ps) = HsPatCons c (map (substPat v1 v2) ps)
+-- | Substitute (in the given alternative) all variables equal to given HsTmVar with second HsTmVar
+substAlt :: Eq a => HsTmVar a -> HsTmVar a -> HsAlt a -> HsAlt a
+substAlt v1 v2 (HsAlt p t) = HsAlt (substPat v1 v2 p) (substTm v1 v2 t)
+-- | Substitute (in the given Term) all variables equal to first HsTmVar with second HsTmVar
+substTm :: Eq a => HsTmVar a -> HsTmVar a -> Term a -> Term a
+substTm v1 v2 t' = case t' of
+  TmVar v
+    | v == v1    -> TmVar    v2
+    | otherwise  -> TmVar    v
+  TmCon c        -> TmCon    c
+  TmAbs v t      -> TmAbs    v (substTm v1 v2 t)
+  TmApp t1 t2    -> TmApp    (substTm v1 v2 t1) (substTm v1 v2 t2)
+  TmLet v t1 t2  -> TmLet    v (substTm v1 v2 t1) (substTm v1 v2 t2)
+  TmCase t as    -> TmCase   (substTm v1 v2 t) (map (substAlt v1 v2) as)
+
+type PmEqn = ([RnPat], RnTerm)
+
+get_rhs :: PmEqn -> RnTerm
+get_rhs = snd
+
+altToEqn :: RnAlt -> PmEqn
+altToEqn (HsAlt p t) = ([p], t)
+
+defaultTerm :: GenM (RnMonoTy, FcTerm)
+defaultTerm = do
+  ty <- TyVar   <$> freshRnTyVar KStar
+  tm <- FcTmVar <$> freshFcTmVar
+  return (ty, tm)
+
+-- | Elaborate a case expression
+elabTmCase :: RnTerm -> [RnAlt] -> GenM (RnMonoTy, FcTerm)
+elabTmCase scr ((HsAlt (HsPatVar x) rhs):_) = elabTmApp (TmAbs x rhs) scr
+elabTmCase scr alts = do
+  let qs           = map altToEqn alts
+  x                <- makeVar
+  def              <- defaultTerm
+  match scr [x] qs def
+
+-- | Check if first pattern of equation contains a variable
+isVar :: PmEqn -> Bool
+isVar (((HsPatVar _):_), _) = True
+isVar _                     = False
+
+-- | Check if first pattern of equation contains a constructor
+isCon :: PmEqn -> Bool
+isCon (((HsPatCons _ _):_), _) = True
+isCon _                        = False
+
+-- | Get the constructor from an alternative equation
+getCon :: PmEqn -> RnDataCon
```

Comment:
I feel like this code would be a lot nicer, returning a Maybe instead of throwing errors.